pkg/sql/rollback.go

```diff
@@ -0,0 +1,226 @@
+//
+// This function will error, without a resume span, if it encounters an intent
+// in the span. The caller should resolve these intents by retrying the
+// function. To prevent errors, the caller should only pass a table that will
+// not see new writes during this bulk delete operation (e.g. on a span that's
+// part of an import rollback).
+//
+// NOTE: this function will issue tombstones on keys with versions that match
+// the predicate, in contrast to RevertRange, which rolls back a table to a
+// certain timestamp. For example, if a key gets a single update after the
+// predicate.startTime, DeleteTableWithPredicate would delete that key, while
+// RevertRange would revert that key to its state before the update.
+func DeleteTableWithPredicate(
+	ctx context.Context,
+	db *kv.DB,
+	codec keys.SQLCodec,
+	distSender *kvcoord.DistSender,
+	table catalog.TableDescriptor,
+	predicates roachpb.DeleteRangePredicates,
+	batchSize int64,
+) error {
+
+	log.Infof(ctx, "deleting data for table %d with predicate %s", table.GetID(), predicates.String())
+	tableKey := roachpb.RKey(codec.TablePrefix(uint32(table.GetID())))
+	tableSpan := roachpb.RSpan{Key: tableKey, EndKey: tableKey.PrefixEnd()}
+
+	// TODO (msbutler): tune these
+	rangesPerBatch := int64(100)
+	numworkers := 4
+
+	initialSpans := make(roachpb.Spans, 1)
+
+	// Partition the inputted span such that each span
+	// partition contains 100 ranges of data.
+	var n int64
+	lastKey := tableSpan.Key
+	ri := kvcoord.MakeRangeIterator(distSender)
+	for ri.Seek(ctx, tableSpan.Key, kvcoord.Ascending); ; ri.Next(ctx) {
+		if !ri.Valid() {
+			return ri.Error()
+		}
+		if n++; n >= rangesPerBatch || !ri.NeedAnother(tableSpan) {
+			endKey := ri.Desc().EndKey
+			if tableSpan.EndKey.Less(endKey) {
+				endKey = tableSpan.EndKey
+			}
+			initialSpans = append(initialSpans, roachpb.Span{Key: lastKey.AsRawKey(),
+				EndKey: endKey.AsRawKey()})
+			n = 0
+			lastKey = endKey
+		}
+
+		if !ri.NeedAnother(tableSpan) {
+			break
+		}
+	}
+
+	spansToDo := make(chan *roachpb.Span, len(initialSpans))
+	for _, span := range initialSpans {
+		spansToDo <- &span
+	}
+
+	// Spin up a few workers to issue DeleteRange requests on the spans in the
+	// spansToDo channel. Using several workers allows the distsender to handle
+	// the span partitions of the initial span concurrently.
+	grp := ctxgroup.WithContext(ctx)
+	grp.GoCtx(func(ctx context.Context) error {
+		return ctxgroup.GroupWorkers(ctx, numworkers, func(ctx context.Context, _ int) error {
+			for {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case span := <-spansToDo:
+					admissionHeader := roachpb.AdmissionHeader{
+						Priority:                 int32(admissionpb.BulkNormalPri),
+						CreateTime:               timeutil.Now().UnixNano(),
+						Source:                   roachpb.AdmissionHeader_FROM_SQL,
+						NoMemoryReservedAtSource: true,
+					}
+					delRangeRequest := &roachpb.DeleteRangeRequest{
```

Comment:
Is it worth asserting that the version gate is active here, or earlier in the function? If it wasn't, we'd be deleting all the data in the range.

Reply:
hmmmm. The caller already does this here and import rollbacks will be the only caller of this funciton.

https://github.com/cockroachdb/cockroach/blob/eb8d649813bd3ded2f80cba340252d553a28b85d/pkg/sql/importer/import_job.go#L1466-L1469

I suppose we could pass a handler to cluster settings to the function, but perhaps it's overkill?

Reply:
Not a big deal, as long as we're reasonably certain that callers won't forget to check the gate.